test/Ocelot.UnitTests/DownstreamRouteFinder/DownstreamRouteFinderTests.cs

```diff
@@ -683,6 +701,114 @@ public void should_return_route_when_host_matches_but_null_host_on_same_path_fir
+                .And(x => x.GivenTheConfigurationIs(new List<Route>
+                    {
+                        new RouteBuilder()
+                            .WithDownstreamRoute(new DownstreamRouteBuilder()
+                                .WithDownstreamPathTemplate("someDownstreamPath")
+                                .WithUpstreamHttpMethod(new List<string> { "Get" })
+                                .WithUpstreamPathTemplate(new UpstreamPathTemplate("someUpstreamPath", 1, false, "someUpstreamPath"))
+                                .Build())
+                            .WithUpstreamHttpMethod(new List<string> { "Get" })
+                            .WithUpstreamPathTemplate(new UpstreamPathTemplate("someUpstreamPath", 1, false, "someUpstreamPath"))
+                            .WithUpstreamHeaderRoutingOptions(_upstreamHeaderRoutingOptions)
+                            .Build(),
+                    }, string.Empty, serviceProviderConfig
+                ))
+                .And(x => x.GivenTheUrlMatcherReturns(new OkResponse<UrlMatch>(new UrlMatch(true))))
+                .And(x => x.GivenTheUpstreamHttpMethodIs("Get"))
+                .And(x => x.GivenEmptyRequestHeaders())
+                .When(x => x.WhenICallTheFinder())
+                .Then(x => x.ThenAnErrorResponseIsReturned())
+                .BDDfy();
+        }
+
+        [Fact]
+        public void should_not_return_route_with_upstream_header_routing_options_enabled_and_non_matching_request_headers()
+        {
+            var serviceProviderConfig = new ServiceProviderConfigurationBuilder().Build();
+
+            this.Given(x => x.GivenThereIsAnUpstreamUrlPath("matchInUrlMatcher/"))
+                .And(x => x.GivenTheTemplateVariableAndNameFinderReturns(
+                        new OkResponse<List<PlaceholderNameAndValue>>(
+                            new List<PlaceholderNameAndValue>())))
+                .And(x => x.GivenUpstreamHeaderRoutingOptions())
+                .And(x => x.GivenTheConfigurationIs(new List<Route>
+                    {
+                        new RouteBuilder()
+                            .WithDownstreamRoute(new DownstreamRouteBuilder()
+                                .WithDownstreamPathTemplate("someDownstreamPath")
+                                .WithUpstreamHttpMethod(new List<string> { "Get" })
+                                .WithUpstreamPathTemplate(new UpstreamPathTemplate("someUpstreamPath", 1, false, "someUpstreamPath"))
+                                .Build())
+                            .WithUpstreamHttpMethod(new List<string> { "Get" })
+                            .WithUpstreamPathTemplate(new UpstreamPathTemplate("someUpstreamPath", 1, false, "someUpstreamPath"))
+                            .WithUpstreamHeaderRoutingOptions(_upstreamHeaderRoutingOptions)
+                            .Build(),
+                    }, string.Empty, serviceProviderConfig
+                ))

```

Comment:
DRY violation!
So, I see copied & pasted code blocks which can be wrapped as private helper.

Reply:
Honestly, I would prefer keeping the code separate. I know parts of it are the same, but it's a unit test and I believe unit tests should be as separate and self-contained as possible.

Reply:
True argument when you said about self-contained.
But it is opposite to readability and focusing of key changes during test setup phase.
It's dilemma!
Nothing fixed, right?

Reply:
No, I have not made any changes in this file since your previous comment.

Reply:
Will help you later...

Reply:
**Fixed** in https://github.com/ThreeMammals/Ocelot/pull/1684/commits/06ff9adc29d7709847a8eeb281597a3df138533d